### PR TITLE
Explain NA gene set scores in the Oncoprint legend

### PR DESCRIPTION
**Pull request by fedde-s**
Mark the objects that represent cells with missing values using the
property that is used for this in the rest of the Oncoprint, to make use
of the N/A legend.

Oncoprint with a gene set legend that explains what the grey cells for one sample mean

### CODE REVIEW

**Comment by fedde-s** [comment]
@adamabeshouse, could you review if this is indeed the solution you suggested? Patches in this branch will eventually be presented as a pull request to [the rc branch of cBioPortal/cbioportal](https://github.com/cBioPortal/cbioportal/commits/rc/), if I'm not mistaken about @pieterlukasse's planning for this.

File: portal/src/main/webapp/js/api/cbioportal-datamanager.js [hunk 1]
```diff
@@ -1144,6 +1147,7 @@ window.initDatamanager = function (genetic_profile_ids, oql_query, geneset_ids,
 		if (data_by_id[geneset_id][case_id].profile_data === null) {
 		    // set the initial value for this sample or patient
 		    data_by_id[geneset_id][case_id].profile_data = sample_score;
+		    data_by_id[geneset_id][case_id].na = false;
```
Comment (adamabeshouse) [inline portal/src/main/webapp/js/api/cbioportal-datamanager.js:1150]: just don't set the property in this case

File: portal/src/main/webapp/js/api/cbioportal-datamanager.js [hunk 2]
```diff
@@ -1049,7 +1049,8 @@ window.initDatamanager = function (genetic_profile_ids, oql_query, geneset_ids,
      * @property {string} study - identifier of the study to which this sample belongs
      * @property {string} sample - identifier of the sample
      * @property {number} uid - cross-study identifier of this sample in the current Oncoprint
-     * @property {number} profile_data - score of the sample or patient for this gene set
+     * @property {?number} profile_data - score of the sample for this gene set
```
Comment (adamabeshouse) [inline portal/src/main/webapp/js/api/cbioportal-datamanager.js]: shouldnt this be number or null, not number or undefined

File: portal/src/main/webapp/js/api/cbioportal-datamanager.js [hunk 3]
```diff
@@ -1059,7 +1060,8 @@ window.initDatamanager = function (genetic_profile_ids, oql_query, geneset_ids,
      * @property {string} study - identifier of the study to which this patient belongs
      * @property {string} patient - identifier of the patient
      * @property {number} uid - cross-study identifier of this patient in the current Oncoprint
-     * @property {number} profile_data - score of the sample or patient for this gene set
+     * @property {?number} profile_data - score of the patient for this gene set
```
Comment (adamabeshouse) [inline portal/src/main/webapp/js/api/cbioportal-datamanager.js]: number or null, not number or undefined?

**Review by adamabeshouse — CHANGES_REQUESTED** [review]

File: portal/src/main/webapp/js/api/cbioportal-datamanager.js (hunk 2 above)
Comment (fedde-s) [reply portal/src/main/webapp/js/api/cbioportal-datamanager.js]: @adamabeshouse  A prefixed question mark is the JSDoc syntactic sugar for ‘nullable’: http://usejsdoc.org/tags-type.html

Since we don't actually use JSDoc to generate API docs, and we had the plan to eventually express this in TypeScript, I think it would be useful if I rephrase this to an explicit union type `(number|null)` to avoid this confusion while thinking in TypeScript.

File: portal/src/main/webapp/js/api/cbioportal-datamanager.js (hunk 3 above)
Comment (fedde-s) [reply portal/src/main/webapp/js/api/cbioportal-datamanager.js]: @adamabeshouse  As stated [above](https://github.com/thehyve/cbioportal/pull/18#discussion_r123955269).

File: portal/src/main/webapp/js/api/cbioportal-datamanager.js (hunk 1 above)
Comment (fedde-s) [reply portal/src/main/webapp/js/api/cbioportal-datamanager.js:1150]: @adamabeshouse  Following the original example in https://github.com/cBioPortal/cbioportal/pull/1847, this loop iterates over the scores received from the API to add them to the appropriate records, and leaves the others untouched at the default of `{profile_data: null, na: true}`. I believe that deleting properties may be less efficient in some browsers than simply assigning. Would you prefer if I assigned `undefined` instead of `false`?

Comment (fedde-s) [reply portal/src/main/webapp/js/api/cbioportal-datamanager.js:1150]: Never mind, I see no evidence of `delete` being inefficient. @adamabeshouse, do you think deleting the property is better than setting it to false, even though it's a boolean?

Comment (fedde-s) [reply portal/src/main/webapp/js/api/cbioportal-datamanager.js:1150]: Oh, https://jsperf.com/delete-vs-undefined-vs-null/3 does show that deleting properties takes nearly twice as long as assigning in the browsers I have access to right now, but considering that might be premature optimisation.

Comment (adamabeshouse) [reply portal/src/main/webapp/js/api/cbioportal-datamanager.js:1150]: oh I understand, sorry I didn't realize the way this was working. this should be right, then. thanks for looking into it

**Review by adamabeshouse — APPROVED** [review]

File: portal/src/main/webapp/js/api/cbioportal-datamanager.js (hunk 1 above)
Comment (fedde-s) [reply portal/src/main/webapp/js/api/cbioportal-datamanager.js:1150]: Nice, thanks for confirming!

**Comment by fedde-s** [comment]
@pieterlukasse, I think this should this be included in the branch and deployed to any servers already testing the gene set functionality then.

**Comment by fedde-s** [comment]
Or I could squash the fixup commit first.

**Comment by fedde-s** [comment]
Squashed. @pieterlukasse, up to you now.

**Review by pieterlukasse — APPROVED** [review]
